vm/src/pyobject.rs

```diff
@@ -829,11 +840,17 @@ impl AttributeProtocol for PyObjectRef {
     }
 }
 
+/// Represents a Python dict
+/// Note: it may also be implemented for Modules and scopes
 pub trait DictProtocol {
+    /// Return whether key is present
     fn contains_key(&self, k: &str) -> bool;
+    /// Gets an item from the dictionary. Returns None if the item is not present.
     fn get_item(&self, k: &str) -> Option<PyObjectRef>;
+    /// Return all the key value pairs present in the dictionary.
     fn get_key_value_pairs(&self) -> Vec<(PyObjectRef, PyObjectRef)>;
-    fn set_item(&self, ctx: &PyContext, key: &str, v: PyObjectRef);
+    /// Sets an item value. This is equivalent to the Python expression `self[key] = value`.
+    fn set_item(&self, key: PyObjectRef, v: PyObjectRef);
```

Comment:
We require here a mutable reference to the virtualmachine, then we can call simply `__setitem__` in the implementation. This way, we can support custom objects with setitem. Furthermore, when we have full dictionary support, we require the virtualmachine to call `__hash__` and `__eq__` magic methods to set an item in the dictionary.

Please note that introducing the VirtualMachine here, requires also modifications to `set_attr` which uses `set_item`. So it might be easier to start with modifying `set_attr`.